src/hotspot/share/jfr/periodic/sampling/jfrCPUTimeThreadSampler.cpp

```diff
@@ -0,0 +1,717 @@
+/*
+ * Copyright (c) 2025 SAP SE. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "jfr/periodic/sampling/jfrCPUTimeThreadSampler.hpp"
+
+
+#if defined(LINUX)
+#include "jfr/periodic/sampling/jfrThreadSampling.hpp"
+#include "jfr/support/jfrThreadLocal.hpp"
+#include "jfr/utilities/jfrTime.hpp"
+#include "jfr/utilities/jfrThreadIterator.hpp"
+#include "jfr/utilities/jfrTypes.hpp"
+#include "jfrfiles/jfrEventClasses.hpp"
+#include "memory/resourceArea.hpp"
+#include "runtime/atomic.hpp"
+#include "runtime/javaThread.hpp"
+#include "runtime/osThread.hpp"
+#include "runtime/safepointMechanism.inline.hpp"
+#include "runtime/threadSMR.hpp"
+#include "runtime/vmOperation.hpp"
+#include "runtime/vmThread.hpp"
+#include "utilities/ticks.hpp"
+
+#include "signals_posix.hpp"
+
+static const int64_t AUTOADAPT_INTERVAL_MS = 100;
+
+static bool is_excluded(JavaThread* jt) {
+  return jt->is_hidden_from_external_view() ||
+         jt->jfr_thread_local()->is_excluded() ||
+         jt->is_JfrRecorder_thread();
+}
+
+static JavaThread* get_java_thread_if_valid() {
+  Thread* raw_thread = Thread::current_or_null_safe();
+  if (raw_thread == nullptr) {
+    // probably while shutting down
+    return nullptr;
+  }
+  assert(raw_thread->is_Java_thread(), "invariant");
+  JavaThread* jt = JavaThread::cast(raw_thread);
+  if (is_excluded(jt) || jt->is_exiting()) {
+    return nullptr;
+  }
+  return jt;
+}
+
+JfrCPUTimeTraceQueue::JfrCPUTimeTraceQueue(u4 capacity) :
+  _capacity(capacity), _head(0), _lost_samples(0) {
+  _data = JfrCHeapObj::new_array<JfrCPUTimeSampleRequest>(capacity);
+}
+
+JfrCPUTimeTraceQueue::~JfrCPUTimeTraceQueue() {
+  JfrCHeapObj::free(_data, _capacity * sizeof(JfrCPUTimeSampleRequest));
+}
+
+bool JfrCPUTimeTraceQueue::enqueue(JfrCPUTimeSampleRequest& request) {
+  assert(JavaThread::current()->jfr_thread_local()->is_cpu_time_jfr_enqueue_locked(), "invariant");
```

Comment:
What is preventing another thread from enqueuing a request here? We only know it holds a thread-local lock? Let's make it explicit at this site that the current queue corresponds to the thread-local queue for the current thread.

+ assert(&JavaThread::current()->jfr_thread_local()->cpu_time_jfr_queue() == this, "invariant");